File: clients/fsharp/Debugging/DebugState.fs
#nowarn "0058"

namespace Komus24
namespace Komus24.Debugging

open Komus24

/// App state for debugging
type DebugState = {
    /// Currently pressed keys
    PressedKeys: string[];
} with

    /// Write DebugState to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.PressedKeys.Length
        this.PressedKeys |> Array.iter (fun value ->
            let valueData : byte[] = System.Text.Encoding.UTF8.GetBytes value
            writer.Write valueData.Length
            writer.Write valueData )
        ()

    /// Read DebugState from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        PressedKeys = [|for _ in 1 .. reader.ReadInt32() do
                          yield reader.ReadInt32() |> reader.ReadBytes |> System.Text.Encoding.UTF8.GetString |]
    }

File: clients/rust/src/debugging/debug_state.rs
use super::*;

/// App state for debugging
#[derive(Clone, Debug)]
pub struct DebugState {
    /// Currently pressed keys
    pub pressed_keys: Vec<String>,
}

impl trans::Trans for DebugState {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.pressed_keys.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let pressed_keys: Vec<String> = trans::Trans::read_from(reader)?;
        Ok(Self {
            pressed_keys,
        })
    }
}